cpp/src/utility/bbox_thrust.cuh

```diff
@@ -0,0 +1,220 @@
+    min(fx1,fx2),min(fy1,fy2),max(tx1,tx2),max(ty1,ty2));*/
+
+    return SBBox<T>(ll, ur);
+  }
+};
+
+template <typename T>
+struct bbox_transformation {
+  __device__ SBBox<T> operator()(const thrust::tuple<T, T> &t) {
+    return SBBox<T>(t, t);
+  }
+};
+
+template <typename T>
+struct bbox2tuple {
+  __device__ thrust::tuple<T, T, T, T> operator()(const SBBox<T> &bbox) {
+    T x1 = thrust::get<0>(bbox.first);
+    T x2 = thrust::get<0>(bbox.second);
+    T y1 = thrust::get<1>(bbox.first);
+    T y2 = thrust::get<1>(bbox.second);
+    // printf("bbox2tuple: %10.5f %10.5f %10.5f %10.5f\n",x1,y1,x2,y2);
+    return thrust::make_tuple(x1, y1, x2, y2);
+  }
+};
+
+template <typename T>
+struct tuple2bbox {
+  __device__ SBBox<T> operator()(thrust::tuple<T, T, T, T> v) {
+    T x1 = thrust::get<0>(v);
+    T y1 = thrust::get<1>(v);
+    T x2 = thrust::get<2>(v);
+    T y2 = thrust::get<3>(v);
+    // printf("tuple2bbox: %10.5f %10.5f %10.5f %10.5f\n",x1,y1,x2,y2);
+    return SBBox<T>(thrust::make_tuple(x1, y1), thrust::make_tuple(x2, y2));
+  }
+};
+
+template <typename T>
+struct pairwise_test_intersection {
+  uint32_t M;
+  uint32_t num_node;
+  const uint32_t *d_p_key = nullptr;
+  const uint8_t *d_p_lev = nullptr;
+  const bool *d_p_qtsign = nullptr;
+  double scale;
+  SBBox<double> aoi_bbox;
+  const SBBox<T> *ply_bbox;
+
+  pairwise_test_intersection(uint32_t _M, uint32_t _num_node,
+                             const SBBox<double> &_aoi_bbox, double _scale,
+                             const uint32_t *_d_p_key, const uint8_t *_d_p_lev,
+                             const bool *_d_p_sign, const SBBox<T> *_ply_bbox)
+      : M(_M),
+        num_node(_num_node),
+        aoi_bbox(_aoi_bbox),
+        d_p_key(_d_p_key),
+        d_p_lev(_d_p_lev),
+        d_p_qtsign(_d_p_sign),
+        scale(_scale),
+        ply_bbox(_ply_bbox) {
+    // printf("pairwise_test_intersection: num_node=%d\n",num_node);
+  }
+
+  __device__ thrust::tuple<uint8_t, uint8_t, uint32_t, uint32_t> operator()(
+      uint32_t p) const {
+    uint32_t poly_idx = p / num_node;
+    uint32_t quad_idx = p % num_node;
+
+    uint8_t lev = d_p_lev[quad_idx];
+    double s = scale * pow(2.0, M - 1 - lev);
+    uint32_t zx = z_order_x(d_p_key[quad_idx]);
+    uint32_t zy = z_order_y(d_p_key[quad_idx]);
+
+    double x0 = thrust::get<0>(aoi_bbox.first);
+    ;
+    double y0 = thrust::get<1>(aoi_bbox.first);
+    double qx1 = zx * s + x0;
+    double qx2 = (zx + 1) * s + x0;
+    double qy1 = zy * s + y0;
+    double qy2 = (zy + 1) * s + y0;
+
+    SBBox<T> box = ply_bbox[poly_idx];
+    double px1 = thrust::get<0>(box.first);
+    double py1 = thrust::get<1>(box.first);
+    double px2 = thrust::get<0>(box.second);
+    double py2 = thrust::get<1>(box.second);
+    bool not_intersect = (qx1 > px2 || qx2 < px1 || qy1 > py2 || qy2 < py1);
+    uint8_t type = not_intersect ? 2 : (d_p_qtsign[quad_idx] ? 1 : 0);
```

Comment:
```suggestion
    SBBox<T> box = ply_bbox[poly_idx];
    bool not_intersect = (qx1 > max_x(box) ||
                          qx2 < min_x(box) ||
                          qy1 > max_y(box) ||
                          qy2 < min_y(box));
```

Reply:
This type isn't used anymore, but I've kept it here in case the other related in-flight PRs use it.